client/network/src/protocol/generic_proto/handler.rs

```diff
@@ -1,27 +1,1054 @@
+					Poll::Ready(Some(msg)) => msg,
+					Poll::Ready(None) | Poll::Pending => break,
+				};
+
+				match message {
+					NotificationsSinkMessage::Notification {
+						protocol_name,
+						message
+					} => {
+						if let Some(pos) = self.out_protocols.iter().position(|(n, _)| *n == protocol_name) {
+							if let Some(substream) = out_substreams[pos].as_mut() {
+								let _ = substream.start_send_unpin(message);
+								continue 'poll_notifs_sink;
+							}
+
+						} else {
+							log::warn!(
+								target: "sub-libp2p",
+								"Tried to send a notification on non-registered protocol: {:?}",
+								protocol_name
+							);
+						}
+					}
+					NotificationsSinkMessage::ForceClose => {
+						return Poll::Ready(
+							ProtocolsHandlerEvent::Close(NotifsHandlerError::SyncNotificationsClogged)
+						);
+					}
+				}
+			}
+		}
+
+		// The legacy substreams are polled only if the state is `Open`. Otherwise, it would be
+		// possible to receive notifications that would need to get silently discarded.
+		if matches!(self.state, State::Open { .. }) {
+			for n in (0..self.legacy_substreams.len()).rev() {
+				let mut substream = self.legacy_substreams.swap_remove(n);
+				let poll_outcome = Pin::new(&mut substream).poll_next(cx);
+				match poll_outcome {
+					Poll::Pending => self.legacy_substreams.push(substream),
+					Poll::Ready(Some(Ok(RegisteredProtocolEvent::Message(message)))) => {
+						self.legacy_substreams.push(substream);
+						return Poll::Ready(ProtocolsHandlerEvent::Custom(
+							NotifsHandlerOut::CustomMessage { message }
+						))
+					},
+					Poll::Ready(Some(Ok(RegisteredProtocolEvent::Clogged))) => {
+						return Poll::Ready(ProtocolsHandlerEvent::Close(
+							NotifsHandlerError::SyncNotificationsClogged
+						))
+					}
+					Poll::Ready(None) | Poll::Ready(Some(Err(_))) => {
+						if matches!(poll_outcome, Poll::Ready(None)) {
+							self.legacy_shutdown.push(substream);
+						}
+
+						if let State::Open { want_closed, .. } = &mut self.state {
+							if !*want_closed {
+								*want_closed = true;
+								return Poll::Ready(ProtocolsHandlerEvent::Custom(
+									NotifsHandlerOut::CloseDesired
+								))
+							}
+						}
+					}
+				}
+			}
+		}
+
+		shutdown_list(&mut self.legacy_shutdown, cx);
+
+		Poll::Pending
+	}
+}
 
-mod group;
-mod legacy;
-mod notif_in;
-mod notif_out;
+/// Given a list of substreams, tries to shut them down. The substreams that have been successfully
+/// shut down are removed from the list.
+fn shutdown_list
+	(list: &mut SmallVec<impl smallvec::Array<Item = RegisteredProtocolSubstream<NegotiatedSubstream>>>,
+	cx: &mut Context)
+{
+	'outer: for n in (0..list.len()).rev() {
+		let mut substream = list.swap_remove(n);
+		loop {
+			match substream.poll_next_unpin(cx) {
+				Poll::Ready(Some(Ok(_))) => {}
+				Poll::Pending => break,
+				Poll::Ready(Some(Err(_))) | Poll::Ready(None) => continue 'outer,
```

Comment:
Would it not be helpful to log the substream error on `debug`?

Reply:
This specific code handles errors on legacy substreams that we want to get rid of. As such, I don't think it's useful specifically here.